### ISSUE TITLE
[Proposal] nix env script + docker OR remove nix from the CI

### ISSUE THREAD

**Issue by ohad** [idris-lang/Idris2#1507]
nix is an impressive package/configuration manager, but if you don't have it installed and the nix continuous-integration tests fail, it is a pain to fix the error. The purpose of this proposal is to collect opinions on proposals to avoid that.

# Proposal 1: a nix container + scripts
Create some container + script that sets up the environment used by the nix CI.
This way, if one's PR failed, one can easily fire up the environment and debug the problem.

# Proposal 2: remove nix from the CI

The learning curve for nix is sufficiently steep that perhaps we should move it to a separate repository, where nix-specialists could troubleshoot such issues.

**Comment by stefan-hoeck** [comment]
Unless we find someone who is willing to keep fixing nix related issues in this repository on a regular basis, we should opt for 2). Same goes - in my opinion - for other backends / operating systems that are hard to maintain.

**Comment by abailly** [comment]
I fail to understand why nix is part of the _Continous Integration_ process of Idris2. The purpose of CI is to ensure that code submitted can be _integrated_ in the existing codebase. How does this impact nix or the distribution of Idris2 package through nix?

**Comment by fabianhjr** [comment]
@abailly, the contributed nix scripts are mostly for setting up a dev environment (Like Asdf `.tool-versions`) and not for packaging.

**Comment by ohad** [comment]
I agree with @fabianhjr: we have nix scripts in the repo, so we should be testing them during CI.
If we go with proposal 2, we should also remove the nix scripts from the repo so that everything on the repo gets tested.

**Comment by abailly** [comment]
That's exactly what I am saying @ohad, thanks for making it clearer.

**Comment by abailly** [comment]
I don't want to play the troll role here. I have started using nix at work a couple months ago and my experience with nix so far is less than optimal, to say the least. Given that a few smart people out there, even some I know personally, are nix enthusiasts and actively promote its use, I assume this is a limitation on my part. 
The only thing I question is whether a modest contributor like myself, who is merely trying to help in moving a PR forward because he feels it has value for himself and others, should be expected to master troubleshooting builds on Mac, Linux, Windows, _and_ Nix. Nix is not an OS: If build a static binary on Linux and run it on NixOS, it will work. I am willing to troubleshoot issues my PRs might introduce on Windows because well, it's part of Idris' mission to be as widely available as possible. But Nix is first and foremost a package manager, one solution among many other possible solutions: We could also package Idris2 with npm, as a debian/rpm, as MacOS pkg or image... These are all fine, but to me those should be handled elsewhere, in their own repositories, using their own checks, and maintained by expert people.

**Comment by fabianhjr** [comment]
>  If we go with proposal 2, we should also remove the nix scripts from the repo so that everything on the repo gets tested.

I disagree, I think the nix expressions (simplified or not), would be valuable to quickly set up a dev environment even if untested by a CI. If anything that would fit your argument that the nix-expressions should not stop non-nix contributors and the burden of keeping those expressions up to date with the dependencies of Idris2 would be on nix-users.

**Comment by fabianhjr** [comment]
Regarding running a docker image, usig Nix official Alpine-based (as opposed to ubuntu/debian based CI):

```
docker run -it nixos/nix:latest
nix-env -iA nixpkgs.git
git clone https://github.com/idris-lang/Idris2.git
cd Idris2
nix-shell (or other nix commands)
```

**Comment by ohad** [comment]
I think this last comment is a good implementation of Proposal 1. If we stick is somewhere in the repo's documentation (perhaps in `CONTRIBUTING.md`, I'd support it and Proposal 1.

**Comment by ohad** [comment]
The above incantation doesn't allow us to use the  `nix flake check` that runs on the CI, so proposal 1 is blocked until we find the incantation that reproduces a closer environment to the CI. Perhaps this will be unblocked by #1509

**Comment by claymager** [comment]
I can't find a docker container with flakes enabled, but as it's currently set up, 'nix-build -A checks.x86_64-linux` in `nixos/nix` runs everything the nix CI does.

**Comment by gallais** [comment]
> I think the nix expressions (simplified or not), would be valuable to
  quickly set up a dev environment even if untested by a CI.

The problem being that if we ship something that does not work, it reflects
badly on the whole project and not just on the lack of nix maintenance.
We have enough pain points already that we don't want to have people struggle
with a broken install script.

If we provide something, it should be tested and kept in check by the CI.

**Comment by claymager** [comment]
I've addressed a few issues in that PR, but there's one more problem that makes me lean towards excising nix from this repository.

When building from `nixpkgs`, idris2 is pinned to the latest release. When building from `idris2-pkgs`, calling `idris2 --version` provides the specific git short hash, as we get in a standard build.

This information is lost when we build from Nix in this repository, because Nix doesn't have build-time access to the git files. So any time until the next release, `idris2 --version` provides `Idris 2, version 0.4.0`. I haven't seen much use of `--version` in the issues, but it is a useful tool that we don't -- and can't -- have access to here.

**Comment by ueXIMcWizBrdyVEciAlSwxP8aaXP70mO** [comment]
I am not a developer on the project, just an outside observer and a casual user of Idris. I also run NixOS as my daily driver, and so can provide some comments on my experience with Nix and its relation with project setup. Please do bear in mind, however, that I am not a Nix expert.

I would like to echo in part what abailly said. Even though Nix is very useful in configuring systems quickly, there are two things that make it very painful to use:
1. when things fail, it can be very difficult to track down the problem quickly (especially for those still learning Nix)
2. the lack of up-to-date documentation: I think flakes are a particularly egregious example of this
I think all parties involved want the installation process to be as smooth as possible: users want to compile without problems, and developers want to set up their environment without problems. Nix can technically handle both of these scenarios, but the underlying language is fast-moving---e.g. flakes are still not fully rolled out---and opaque---no type signatures, out-of-date documentation---enough to where I think it requires dedicated effort to keep systems functional (no pun intended).

For these reasons, if the CI is to continue relying on Nix for testing, it may be easier wrap the scripts into a container. Including solely the Nix testing expressions in the repository only seems reasonable if everyone were using Nix, which is not the case.
